### PR TITLE
kubeadm: reimplement IPVS check

### LINKED ISSUE #975: RequiredIPVSKernelModulesAvailable warning seems confusing
## What keywords did you search in kubeadm issues before filing this one?
ip_vs, RequiredIPVSKernelModulesAvailable

## Is this a BUG REPORT or FEATURE REQUEST?
BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`):
kubeadm version: &version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:14:41Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:17:28Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:08:34Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
- **Cloud provider or hardware configuration**: 
aws ec2
- **OS** (e.g. from /etc/os-release): 
debian stretch
- **Kernel** (e.g. `uname -a`): 
Linux debian 4.9.0-6-amd64  SMP Debian 4.9.88-1+deb9u1 (2018-05-07) x86_64 GNU/Linux
- **Others**:

## What happened?
ipvs is installed but modules not loaded yet, kubeadm show warning **RequiredIPVSKernelModulesAvailable**
```
# lsmod | grep ip_vs
$ cat kubeadm.yml
kubeProxy:
  config:
    mode: ipvs
    ipvs:
      scheduler: lc
$ sudo kubeadm init --config kubeadm.yml
...
I0706 07:18:56.609843    1084 kernel_validator.go:96] Validating kernel config
	[WARNING RequiredIPVSKernelModulesAvailable]: the IPVS proxier will not be used, because the following required kernel modules are not loaded: [ip_vs_sh ip_vs ip_vs_rr ip_vs_wrr] or no builtin kernel ipvs support: map[ip_vs_wrr:{} ip_vs_sh:{} nf_conntrack_ipv4:{} ip_vs:{} ip_vs_rr:{}]
you can solve this problem with following methods:
 1. Run 'modprobe -- ' to load missing kernel modules;
2. Provide the missing builtin kernel ipvs support

 Pulling images required for setting up a Kubernetes cluster
...
```
But after cluster is up, ipvs modules are actually loaded by kube-proxy and i see Services VIPs in `ipvsadm -ln`

## What you expected to happen?
This is confusing.
Maybe warning should only be shown when such modules are not exist in system?

### PR DESCRIPTION
**Pull request by bart0sh**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Used existing IPVS Proxier API CanUseIPVSProxier instead
of custom implementation.

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#975

**Special notes for your reviewer**:

This is yet another attempt to correctly fix kubernetes/kubeadm#975 and get rid of huge confusing warning message that scares users.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — APPROVED** [review]
@bart0sh 
thanks for the cleanup. the change LGTM.
please mind that code freeze is tomorrow..

"verify" needs to pass (there is an import-boss issue).

/approve
/priority important-longterm

**Comment by bart0sh** [comment]
@neolit123 Thank you for the review.

Do you have any idea why import-boss is failing/how to reproduce this?

pull-kubernetes-e2e-gce-100-performance failure seems to be not related to this PR. Some testing infra setup issue I guess.

**Comment by neolit123** [comment]
@bart0sh 
this ticket has some info what import boss is and how to test locally:
https://github.com/kubernetes/kubernetes/issues/68201#issue-356613073

on a quick look it seems not happy with the new dependency from pkg.
you might have to add it here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/.import-restrictions#L51

**Comment by bart0sh** [comment]
@neolit123 thank you for the suggestion! That was exactly it - import-boss was not happy about extra dependencies to the bunch of network-related packages.

**Comment by neolit123** [comment]
@bart0sh 
this PR came rather late and today is code freeze (in roughly 12 hours), so you need to convince the IPVS maintainers to approve.

**Review by cmluciano — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/preflight/checks_darwin.go
Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks_darwin.go]: Unsupported check should not return an error, I believe.
You can look at other unsupported check in the same file: IsDockerSystemdCheck. Systemd doesn't exist on MacOS(the same as IPVSProxier), so check is just returns nil, nil.

File: cmd/kubeadm/app/preflight/checks.go
Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks.go]: Done. Thank you for the review!

File: cmd/kubeadm/app/preflight/checks.go
Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks.go]: Done

**Comment by bart0sh** [comment]
@krousey, @detiber Kindly ping for review/approval.

File: cmd/kubeadm/app/preflight/checks_darwin.go
Comment (detiber) [reply cmd/kubeadm/app/preflight/checks_darwin.go]: @bart0sh is correct, unsupported checks do not return an error, if they did then the check would fail for Mac or Windows and kubeadm would be unable to proceed without the user adding command line options to explicitly ignore the check.

File: cmd/kubeadm/app/preflight/checks_linux.go
Comment (detiber) [inline cmd/kubeadm/app/preflight/checks_linux.go]: Since we are ignoring the other returned value this could be:
```go
if _, err := ipvs.CanUseIPVSProxier(kernelHandler, ipsetInterface); err != nil {
    return nil, append(errors, err)
}
```

**Review by detiber — COMMENTED** [review]
minor nit, otherwise lgtm

File: cmd/kubeadm/app/preflight/checks_linux.go
Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks_linux.go]: @detiber Thank you for the review! I've updated the PR.

File: cmd/kubeadm/app/preflight/checks_darwin.go
Comment (cmluciano) [reply cmd/kubeadm/app/preflight/checks_darwin.go]: I'm still a bit confused on why the Check should not return even a warning on a platform that did not support it. Perhaps the warning/error is exposed elsewhere since I don't believe IPVS will work if we try to leverage kubeadm to install this on MacOS.

Comment (neolit123) [reply cmd/kubeadm/app/preflight/checks_darwin.go]: @cmluciano this is a standard pattern in a variety of programming languages, golang including.
you have separate OS files where if a certain functionality does not work, it is a no-op and does not throw any errors.

**Comment by bart0sh** [comment]
Assigning to approvers for final approval

/assign @thokin @m1093782566 @timothysc

**Comment by bart0sh** [comment]
@m1093782566 @timothysc Kindly ping for approval.

**Review by neolit123 — COMMENTED** [review]
/approve
SGTM on the kubeadm side.

File: cmd/kubeadm/app/preflight/checks_linux.go
Comment (timothysc) [inline cmd/kubeadm/app/preflight/checks_linux.go]: It doesn't look like we return warnings in any way, can we remove the return argument?

**Review by timothysc — COMMENTED** [review]
thanks for the patch, but I have a comment that I'd like an answer to before lgtm 
/approve 
/hold

File: cmd/kubeadm/app/preflight/checks_linux.go
Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks_linux.go]: I don't think we can as this is a part of [Checker interface](https://github.com/kubernetes/kubernetes/blob/18ac575d9be27dbfbd464557b85615bbabfa69ba/cmd/kubeadm/app/preflight/checks.go#L88:L91). If we do that it would break kubeamd build I believe.

**Comment by rosti** [comment]
I think, that this is ready for merge now. Thanks @bart0sh !
/hold cancel

**Pull request merged**
